ember-core/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala

```diff
@@ -57,46 +62,55 @@ private[ember] object ChunkedEncoding {
                         s"Failed to parse chunked header : ${hdr.decodeUtf8}"))
                   case Some(0) =>
                     // Done With Message, Now Parse Trailers
-                    parseTrailers[F](maxChunkHeaderSize)(Stream.chunk(Chunk.byteVector(rem)) ++ tl)
-                      .flatMap { hdrs =>
-                        Pull.eval(trailers.complete(hdrs)) >> Pull.done
-                      }
-                  case Some(sz) => go(Right(sz), Stream.chunk(Chunk.ByteVectorChunk(rem)) ++ tl)
+                    Pull.eval(
+                      parseTrailers[F](maxChunkHeaderSize)(rem.toArray, read)
+                        .flatMap { case (hdrs, bytes) =>
+                          trailers.complete(hdrs) >> rest.set(Some(bytes))
+                        }
+                    ) >> Pull.done
+                  case Some(sz) => go(Right(sz), rem.toArray)
                 }
               }
 
             case Right(remains) =>
               if (remains == bv.size)
-                Pull.output(Chunk.ByteVectorChunk(bv)) >> go(Left(ByteVector.empty), tl)
+                Pull.output(Chunk.ByteVectorChunk(bv)) >> go(
+                  Left(ByteVector.empty),
+                  Array.emptyByteArray)
               else if (remains > bv.size)
-                Pull.output(Chunk.ByteVectorChunk(bv)) >> go(Right(remains - bv.size), tl)
+                Pull.output(Chunk.ByteVectorChunk(bv)) >> go(
+                  Right(remains - bv.size),
+                  Array.emptyByteArray)
               else {
                 val (out, next) = bv.splitAt(remains.toLong)
-                Pull.output(Chunk.ByteVectorChunk(out)) >> go(
-                  Left(ByteVector.empty),
-                  Stream.chunk(Chunk.ByteVectorChunk(next)) ++ tl)
+                Pull.output(Chunk.ByteVectorChunk(out)) >> go(Left(ByteVector.empty), next.toArray)
               }
           }
       }
+    }
 
-    go(Left(ByteVector.empty), _).stream
+    go(Left(ByteVector.empty), head).stream
   }
 
   private def parseTrailers[F[_]: MonadThrow](
       maxHeaderSize: Int
-  )(s: Stream[F, Byte]): Pull[F, Nothing, Headers] =
-    s.pull.uncons.flatMap {
-      case None => Pull.pure(Headers.empty)
-      case Some((chunk, tl)) =>
-        if (chunk.isEmpty) parseTrailers(maxHeaderSize)(tl)
+  )(head: Array[Byte], read: F[Option[Chunk[Byte]]]): F[(Headers, Array[Byte])] = {
+    val uncons = if (head.nonEmpty) (Some(Chunk.bytes(head)): Option[Chunk[Byte]]).pure[F] else read
+    uncons.flatMap {
+      case None => (Headers.empty, Array.emptyByteArray).pure[F]
+      case Some(chunk) =>
+        // TODO: Is this state eliminated because we check for nonemptiness already? Or can we get an empty chunk from read
```

Comment:
We just eliminated some empty chunk logic in Multipart, but only in FS2-3.  It didn't work in FS2-2.